mp4parse_capi/src/lib.rs

```diff
@@ -385,61 +391,95 @@ impl Read for Mp4parseIo {
 
 // C API wrapper functions.
 
-/// Allocate an `Mp4parseParser*` to read from the supplied `Mp4parseIo`.
+/// Allocate an `Mp4parseParser*` to read from the supplied `Mp4parseIo` and
+/// parse the content from the `Mp4parseIo` argument until EOF or error.
 ///
 /// # Safety
 ///
 /// This function is unsafe because it dereferences the io pointer given to it.
 /// The caller should ensure that the `Mp4ParseIo` struct passed in is a valid
 /// pointer. The caller should also ensure the members of io are valid: the
 /// `read` function should be sanely implemented, and the `userdata` pointer
-/// should be valid.
+/// should be valid. The `status_out` is also dereferenced and must point to
+/// a valid Mp4parseStatus.
+///
+/// On a successful the variable pointed to by `status_out` will be set to
+/// `Mp4parseStatus::Ok` and the return value will be a non-NULL pointer to
+/// an `Mp4parseParser`. If there is any sort of error, the return value will
+/// be NULL and the type of error will be set in the variable pointed to by
+/// `status_out`, ensuring the consumer never has access to potentially invalid
+/// parsed data.
+///
+/// To avoid leaking memory, any successful return of this function must be
+/// paired with a call to `mp4_parse_free`. In the event of error, no memory
+/// will be allocated and `mp4_parse_free` must *not* be called.
 #[no_mangle]
-pub unsafe extern fn mp4parse_new(io: *const Mp4parseIo) -> *mut Mp4parseParser {
-    if io.is_null() || (*io).userdata.is_null() {
-        return std::ptr::null_mut();
-    }
-    if (*io).read.is_none() {
-        return std::ptr::null_mut();
-    }
-    let parser = Box::new(Mp4parseParser {
-        context: MediaContext::new(),
-        io: (*io).clone(),
-        poisoned: false,
-        opus_header: HashMap::new(),
-        pssh_data: Vec::new(),
-        sample_table: HashMap::new(),
-        audio_track_sample_descriptions: HashMap::new(),
-        video_track_sample_descriptions: HashMap::new(),
-    });
-
-    Box::into_raw(parser)
+pub unsafe extern fn mp4parse_new(io: *const Mp4parseIo, status_out: *mut Mp4parseStatus) -> *mut Mp4parseParser {
```

Comment:
Any reason not to return Mp4parseStatus and make Mp4parseParser the out parameter?

Reply:
That was what I tried first. With this signature:
```rust
pub unsafe extern fn mp4parse_new(io: *const Mp4parseIo, parser_out: *mut *const Mp4parseParser) -> Mp4parseStatus {
```
It led to some very awkward and error-prone code on the test side. My first attempt (that compiled) looked like this:
```rust
let mut parser: *mut Mp4parseParser = std::ptr::null_mut();
let rv = mp4parse_new(&io, &mut (parser as *const _));
```
But is subtly wrong (can you see the bug?). The "correct" approach is this mess:
```rust
let rv = mp4parse_new(&io, &mut parser as *mut *mut _ as *mut *const _);
```
In the end, it seemed worth switching to a slightly weirder signature, that had this much more ergonomic code for consumers:
```rust
        let mut rv = Mp4parseStatus::Invalid;
        let parser = mp4parse_new(&io, &mut rv);
```
Let me know if you disagree.

Reply:
No problem, I don't think. I guess I was being a bit of a stickler in my first attempt by making the `parser` out parameter `*mut *const` rather than `*mut *mut` to indicate that the caller passes in a pointer to be set, but that the dereferenced value should never be modified. Given that the other accessors *do* require mutation, for example:
```rust
pub unsafe extern fn mp4parse_get_track_info(parser: *mut Mp4parseParser, track_index: u32, info: *mut Mp4parseTrackInfo) -> Mp4parseStatus;
```
it's probably not worth it to make that distinction in `mp4parse_new`.

Let me try this other approach a bit and see if anything particularly ugly pops out. If not, agree with you that having the status be the return value is nicer since it's more consistent with the rest of the API.